### PR TITLE
Initial Conditions

### PR DESCRIPTION
**Pull request by jhdark**
## Proposed changes

Initial conditions! With this change, users will be able to define an initial condition for their simulation.

The value of the initial condition can be a float, int or callable. Although the callable function cannot be a function of time.

Usage:
```python
vol_subdomain = F.VolumeSubdomain1D(1, borders=[0, 1], material=dummy_mat)
my_species = F.Species("A")

init_cond = F.InitialCondition(
        value=1.0, species=my_species
    )
```

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: festim/initial_condition.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+        if isinstance(self.value, (int, float)):
+            self.species.solution.interpolate(
+                F.as_fenics_constant(mesh=mesh, value=self.value)
+            )
+
+        elif callable(self.value):
+            arguments = self.value.__code__.co_varnames
+
+            if "t" in arguments and "x" not in arguments and "T" not in arguments:
```
Comment (RemDelaporteMathurin) [inline festim/initial_condition.py:84]: How come "t" can be in the arguments for an initial condition?

File: festim/initial_condition.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        value (float, int, fem.Constant or callable): the value of the initial condition
+        species (festim.Species): the species to which the source is applied
+
+    Usage:
+        >>> from festim import InitialCondition
+        >>> InitialCondition(volume=my_vol, value=1, species="H")
+        >>> InitialCondition(volume=my_vol, value=lambda x: 1 + x[0], species="H")
+        >>> InitialCondition(volume=my_vol, value=lambda t: 1 + t, species="H")
+        >>> InitialCondition(volume=my_vol, value=lambda T: 1 + T, species="H")
+        >>> InitialCondition(volume=my_vol, value=lambda x, t: 1 + x[0] + t, species="H")
```
Comment (RemDelaporteMathurin) [inline festim/initial_condition.py:25]: According to the docstrings, the `species` argument is of type `festim.Species` not `str`

File: festim/initial_condition.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    """
+
+    def __init__(self, value, volume, species):
+        self.value = value
+        self.volume = volume
```
[the remark this answers is not in the document]
Comment (RemDelaporteMathurin) [reply festim/initial_condition.py:30]: Maybe it's best to keep this for later but the main question here is: how to apply an initial condition in one subdomain only?

File: festim/hydrogen_transport_problem.py [hunk 4]
```diff
@@ -88,6 +88,7 @@ def __init__(
         species=[],
         temperature=None,
         sources=[],
+        initial_conditions=[],
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:91]: This should be documented

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
First round of comments

File: festim/initial_condition.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        x = ufl.SpatialCoordinate(mesh)
+
+        if isinstance(self.value, (int, float)):
+            self.species.prev_solution.x.array[:] = float(self.value)
+
+        elif callable(self.value):
+            arguments = self.value.__code__.co_varnames
+            kwargs = {}
+            if "t" in arguments:
+                raise ValueError("Initial condition cannot be a function of time.")
```
[the remark this answers is not in the document]
Comment (jhdark) [reply festim/initial_condition.py:53]: Can do

File: festim/hydrogen_transport_problem.py [hunk 6]
```diff
@@ -527,6 +535,42 @@ def create_source_values_fenics(self):
...
+                temperature=self.temperature_fenics,
+                function_space=function_space_value,
+            )
+
+            # assign to previous solution of species
+            if not self.multispecies:
+                condition.species.prev_solution.interpolate(condition.expr_fenics)
+            else:
+                for idx, spe in enumerate(self.species):
+                    if spe == condition.species:
```
[the remark this answers is not in the document]
Comment (RemDelaporteMathurin) [reply festim/hydrogen_transport_problem.py:570]: @jhdark can we address this?

It should look something like:

```python
idx = self.species.index(spe)
```

File: test/test_h_transport_problem.py [hunk 7]
```diff
@@ -645,3 +645,101 @@ def test_species_setter():
...
+        ],
+        settings=F.Settings(atol=1, rtol=1, final_time=2, stepsize=1),
+    )
+
+    # RUN
+    my_model.initialise()
+
+    # TEST
+    assert np.isclose(my_model.u_n.sub(0).x.array[-2], expected_value_1)
+    assert np.isclose(my_model.u_n.sub(1).x.array[-1], expected_value_2)
```
Comment (RemDelaporteMathurin) [inline test/test_h_transport_problem.py:745]: It isn't clear to me why we have [-2] and [-1] here

File: festim/initial_condition.py [hunk 8]
```diff
@@ -1,63 +1,67 @@
...
-            self.species.prev_solution.x.array[:] = float(self.value)
+            self.expr_fenics = lambda x: np.full(x.shape[1], self.value)
 
         elif callable(self.value):
             arguments = self.value.__code__.co_varnames
             kwargs = {}
             if "t" in arguments:
-                raise ValueError("Initial condition cannot be a function of time.")
+                warnings.warn("Initial condition cannot be a function of time.")
+                kwargs["t"] = t
```
[the remark this answers is not in the document]
Comment (RemDelaporteMathurin) [reply festim/initial_condition.py:58]: @jhdark can we address this?

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]

File: festim/initial_condition.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jhdark) [reply festim/initial_condition.py:58]: As in we should have `kwargs["t"] = 0`?

[the remark this answers is not in the document]
Comment (RemDelaporteMathurin) [reply festim/initial_condition.py:58]: yes

**Comment by RemDelaporteMathurin** [comment]
I don't see where it uses `InitialCondition.subdomain` are you sure that initial values in subdomains is implemented?

File: test/test_h_transport_problem.py [hunk 9]
```diff
@@ -688,6 +745,46 @@ def test_create_species_from_trap():
...
+        ],
+        settings=F.Settings(atol=1, rtol=1, final_time=2, stepsize=1),
+    )
+
+    # RUN
+    my_model.initialise()
+
+    # TEST
+    assert np.isclose(my_model.u_n.x.array[-2], expected_value_1)
+    assert np.isclose(my_model.u_n.x.array[-1], expected_value_2)
```
Comment (RemDelaporteMathurin) [inline test/test_h_transport_problem.py:785]: Could we add just a comment here explaining the [-2] [-1] thing?

**Pull request merged**